cli/command/completion/functions.go

```diff
@@ -0,0 +1,94 @@
+package completion
+
+import (
+	"os"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/command/formatter"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/spf13/cobra"
+)
+
+// ValidArgsFn a function to be used by cobra command as `ValidArgsFunction` to offer command line completion
+type ValidArgsFn func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective)
+
+// CompleteImageNames offers completion for images present within the local store
+func CompleteImageNames(dockerCli command.Cli) ValidArgsFn {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		list, err := dockerCli.Client().ImageList(cmd.Context(), types.ImageListOptions{})
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveError
+		}
+		var names []string
+		for _, image := range list {
+			names = append(names, image.RepoTags...)
+		}
+		return names, cobra.ShellCompDirectiveNoFileComp
+	}
+}
+
+// CompleteContainerNames offers completion for container names and IDs
+// By default, only names are returned.
+// Set DOCKER_COMPLETION_SHOW_CONTAINER_IDS=yes to also complete IDs.
+func CompleteContainerNames(dockerCli command.Cli, all bool, filters ...func(types.Container) bool) ValidArgsFn {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		list, err := dockerCli.Client().ContainerList(cmd.Context(), types.ContainerListOptions{
+			All: all,
+		})
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveError
+		}
+
+		showContainerIDs := os.Getenv("DOCKER_COMPLETION_SHOW_CONTAINER_IDS") == "yes"
+
+		var names []string
+		for _, container := range list {
+			ok := true
+			for _, fn := range filters {
+				if !fn(container) {
+					ok = false
+					break
+				}
+			}
+			if !ok {
+				continue
+			}
```

Comment:
Perhaps more readable (the double negative is a bit hard to read); reverse this and to rename the variable to `skip` (e.g.);

```suggestion
			skip := false
			for _, fn := range filters {
				if !fn(container) {
					skip = true
					break
				}
			}
			if skip {
				continue
			}
```